asst2/CODE/memory_finder.c

```diff
@@ -0,0 +1,438 @@
+#include <signal.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include "my_malloc.h"
+#include "global_vals.h"
+#include "memory_finder.h"
+#include "direct_mapping.h"
+#include "open_address_ht.h"
+
+/**
+ * Swaps data from page at indexA to page at indexB, updates page table and hash table to match the change
+ * @param indexA		first page index to be swapped
+ * @param indexB		second page index to be swapped
+ */
+void swap(int indexA, int indexB){
+	if (indexA == indexB) {return;}
+	pagedata* pdA = (pagedata*)myblock+indexA;
+	pagedata* pdB = (pagedata*)myblock+indexB;
+	ht_put(ht_space, pdA->pid, (ht_key)pg_index(pdA), indexB);
+	ht_put(ht_space, pdB->pid, (ht_key)pg_index(pdB), indexA);
+	pagedata pdTemp;
+	pg_write_pagedata(&pdTemp, pdA->pid, pg_block_occupied(pdA), pg_is_overflow(pdA), pg_index(pdA), pdA->length);
+	pg_write_pagedata(pdA, pdB->pid, pg_block_occupied(pdB), pg_is_overflow(pdB), pg_index(pdB), pdB->length);
+	pg_write_pagedata(pdB, pdTemp.pid, pg_block_occupied(&pdTemp), pg_is_overflow(&pdTemp), pg_index(&pdTemp), pdTemp.length);
+	char* pageA = (char*)mem_space + indexA*page_size;
+	char* pageB = (char*)mem_space + indexB*page_size;
+	char  pageTemp;
+	size_t i;
+	for (i=0; i<page_size; i++) {
+		pageTemp = pageA[i];
+		pageA[i] = pageB[i];
+		pageB[i] = pageTemp;
+	}
+}
+
+/**
+ * Finds the start and end of curr_id's view of memory, first free page, and number of free pages
+ * @param curr_id		ID of process in question.
+ * @param free_page		pointer to store index of first free page
+ * @param num_free 		pointer to store number of free pages available
+ * @param first_page_index	pointer to store the first page index according to process' view
+ * @param last_page_index	pointer to store the last page index according to process' view
+ */
+int find_contig_space(uint32_t curr_id, int* free_page, int* num_free_pages, int* first_page_index, int* last_page_index){
+	*free_page = -1;
+	*num_free_pages = 0;
+	*first_page_index = INT16_MAX;
+	*last_page_index = -1;
+	for (int index = 0; index < num_pages; index++) {
```

Comment:
is there a better way to do this?

Reply:
maybe store start of the contiguous memory space in pagedata ?

Reply:
ok, just wondering

Reply:
if this is meant to assume our thread memory will always be physically contiguous, we're gonna have a bad time

Reply:
no it just traverses `myblock` to find any pages that are belonging to current process, and then looks at where the process thinks that page is, based on where process thinks pages are we can get a start and end index for where the process' view of its memory space which we can use as a counter/index and as a map to traverse the actual indices of process' pages (using the hashtable) to service malloc/free requests